### DISCUSSION TITLE
Memory Allocations and Garbage Collection

### DISCUSSION THREAD

**Discussion by mahrud** [Macaulay2/M2#1734] [General]
I'm curious about the nuances of memory allocation (malloc) in Macaulay2, at the engine, interpreter, and top level, and I'm hoping to use this space to share insights, best practices, etc.

I'll start off with some questions.

I. Engine
  1. what's the difference between different malloc libraries? I know of:
    - [malloc](https://sourceware.org/glibc/wiki/MallocInternals) from GNU C
    - [jemalloc](http://jemalloc.net/) used [in Rust](https://smallcultfollowing.com/babysteps/blog/2014/11/14/allocators-in-rust/)
    - [tcmalloc](https://gperftools.github.io/gperftools/tcmalloc.html)
    - [malloc substitute](https://github.com/ivmai/bdwgc#general-description) from bdw-gc
    - [some more](https://en.wikipedia.org/wiki/C_dynamic_memory_allocation#Implementations)
  2. are there examples of other programs that have something like `our_new_delete`?
  3. is it possible to hot-swap the malloc library that engine uses with a different one for comparison? (relevant: [replacing malloc in C](https://www.gnu.org/software/libc/manual/html_node/Replacing-malloc.html) and [in Rust](https://doc.rust-lang.org/edition-guide/rust-2018/platform-and-target-support/global-allocators.html))
  4. is it possible to hot-swap the malloc library that bdw-gc uses?

II. Interpreter
  1. has there been an audit of the interpreter's memory usage?
  2. can we know whether the gc is doing easily avoidable work collecting stuff that the interpreter leaves around?
  3. can we know which functions run without any dynamic allocations and which don't?
  3. (probably more to add later)

III. Top level
  1. is it possible to detect memory allocation at the top level?
  2. what are the malloc calls associated to calling a function?
  3. where does tail-end optimization happen? are there other similar optimizations?
  4. what's causing the issue in #488?

**Comment by mahrud** [comment]
Addendum: some interesting references about garbage collectors:
  - [High-performance GC for C++](https://v8.dev/blog/high-performance-cpp-gc)
  - [Oilpan](https://docs.google.com/document/d/1y7_0ni0E_kxvrah-QtnreMlzCDKN3QP4BN1Aw7eSLfY/edit#)
  - [Cross-Component GC](https://research.google/pubs/pub47359/)
  - [V8 GC](https://v8.dev/blog/trash-talk)
  - [memory leaks](https://v8.dev/docs/memory-leaks)
  - [Julia GC](https://docs.julialang.org/en/v1/devdocs/object/)

**Comment by mahrud** [comment]
An interesting feature of Julia's GC:
```julia
julia> @time G = GroebnerBasis.f4(I);
  0.523836 seconds (676.75 k allocations: 39.873 MiB, 7.45% gc time, 77.35% compilation time)
```

**Comment by jkyang92** [comment]
Here's some cool stuff with heap profiling that GHC (haskell) uses https://downloads.haskell.org/~ghc/latest/docs/html/users_guide/profiling.html#profiling-memory-usage. In particular it can plot memory usage by type or by source as a function of time.
